chapters/2-abstraction.md

```diff
@@ -0,0 +1,100 @@
+# Abstraction
+
+Functional abstraction allows us to treat functions as black boxes.
+We just care about what a function does and how to use it.
+
+This is helpful when implementing more complex programs.
+We use **Wishful Thinking** to imagine that we have functions
+which do what we want. 
+We start by thinking of what they should do and how to use them.
+This helps you to think about what functions and data you need
+before you implement the function.
+
+## Writing Stubs
+
+When writing a program, we can practise wishful thinking by writing
+**function stubs**.
+
+A [function stub](https://en.wikipedia.org/wiki/Method_stub)
+is a piece of code that stands in for the actual functionality.
+
+The stub allows us to get the main outline of the program up
+before figuring out the details.
+While incrementally working towards the final product,
+the program can be run without syntax errors, allowing us to write
+small manageable chunks at a time.
+
+## Defeating the Balrog
+
+We had quite a few questions about the Balrog lecture training question.
+The main difficulties were not knowing what the question wanted,
+and unfamiliarity with Python's syntax for defining functions.
+
+### Similar Examples
+
+A closer examination of the code given in the question reveals
+two stubs: `spawn_balrog` and `balrog_attack`.
+
+We'll cover their main parts below.
+
+#### Function Definition
+
+Both functions start with the Python keyword `def`, for "define",
+followed by the function's name. Finally, variable names
+for the function's parameters are surrounded by parentheses.
+
+#### Docstrings
+
+Notice that both functions have an English sentence right below the `def`
+line. This is a **docstring**. It's meant to document your code,
+explaining what the function is supposed to do.
+
+[PEP 257](https://www.python.org/dev/peps/pep-0257/) explains Python's
+docstring conventions.
+
+*Why don't we just use a comment?*
+
+Docstrings are useful for documentation generators such as
```

Comment:
maybe can tell them that the string will be shown when they do `help(thier_function)`. More relevant.